### PR TITLE
util: skip always false DNF

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/40997

Problem Summary:
For example:
```sql
explain
select
  *
from
  t71706696 force index(dt_2)
where
  (
    (
      dt = '20210112'
      and db_id = '62812'
      and tbl_id > '228892694'
    )
    or (
      dt = '20210112'
      and db_id = '62813'
      and tbl_id <= '226785696'
    )
    or (
      dt = '20210112'
      and db_id > '62812'
      and db_id < '62813'
    )
  )
```

In the where clause we have CNF contains some DNF and the last DNF is always false. But it will cause the optimizer to fall back to a full table scan.

The final plan looks like:
```sql
+-------------------------------+----------+-----------+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                            | estRows  | task      | access object                                  | operator info                                                                                                                                                                                |
+-------------------------------+----------+-----------+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| IndexLookUp_8                 | 0.67     | root      |                                                |                                                                                                                                                                                |
| ├─Selection_7(Build)          | 0.67     | cop[tikv] |                                                | or(and(eq(test.t71706696.dt, "20210112"), and(eq(test.t71706696.db_id, 62812), gt(test.t71706696.tbl_id, 228892694))), or(and(eq(test.t71706696.dt, "20210112"), and(eq(test.t71706696.db_id, 62813), le(test.t71706696.tbl_id, 226785696))), and(eq(test.t71706696.dt, "20210112"), and(gt(test.t71706696.db_id, 62812), lt(test.t71706696.db_id, 62813))))) |
| │ └─IndexFullScan_5           | 10000.00 | cop[tikv] | table:t71706696, index:dt_2(dt, db_id, tbl_id) | keep order:false, stats:pseudo                                                                                                                                                                                |
| └─TableRowIDScan_6(Probe)     | 0.67     | cop[tikv] | table:t71706696                                | keep order:false, stats:pseudo                                                                                                                                                                                |
+-------------------------------+----------+-----------+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Actually, we can ignore this always false filter and do an index scan.

### What changed and how does it work?

From the code details, we should ignore the empty range here:
https://github.com/pingcap/tidb/blob/3f2468115cf4089d1bf47da3da4d06ca2dd31850/pkg/util/ranger/detacher.go#L739

In the `detachCNFCondAndBuildRangeForIndex` we will return directly from here if the range is empty:
https://github.com/pingcap/tidb/blob/3f2468115cf4089d1bf47da3da4d06ca2dd31850/pkg/util/ranger/detacher.go#L315

But we didn't consider it in the `detachDNFCondAndBuildRangeForIndex`.

So I skip it here:
https://github.com/hi-rustin/tidb/blob/3f2468115cf4089d1bf47da3da4d06ca2dd31850/pkg/util/ranger/detacher.go#L740

```go
			// If CNF item always false, we can return ignore this CNF item.
			if len(ranges) == 0 {
				continue
			}
```

I also updated the `TestIndexMergeIssue49605` test case. Because in this case we have an empty range: `BETWEEN _UTF8MB4'WVz' AND _UTF8MB4'G#駧褉ZC領*lov'`. So that it doesn't use index merge anymore.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
增强 TiDB 对总是 false 的 DNF 的处理，直接忽略这种 filter
Enhance TiDB's handling of DNFs that are always false by simply ignoring such filters
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: pkg/executor/explainfor_test.go [hunk 1]
```diff
@@ -379,15 +379,15 @@ func TestIssue28259(t *testing.T) {
...
 	res = tk.MustQuery("explain for connection " + strconv.FormatUint(tkProcess.ID, 10))
-	require.Len(t, res.Rows(), 3)
-	require.Regexp(t, ".*Selection.*", res.Rows()[1][0])
-	require.Regexp(t, ".*IndexFullScan.*", res.Rows()[2][0])
+	require.Len(t, res.Rows(), 2)
+	require.Regexp(t, ".*IndexReader.*", res.Rows()[0][0])
+	require.Regexp(t, ".*IndexRangeScan.*", res.Rows()[1][0])
 
 	res = tk.MustQuery("explain format = 'brief' select col1 from UK_GCOL_VIRTUAL_18588 use index(UK_COL1) " +
 		"where col1 between -1696020282760139948 and -2619168038882941276 or col1 < -4004648990067362699;")
```
Comment (hi-rustin) [inline pkg/executor/explainfor_test.go:387]: -1696020282760139948 > -2619168038882941276. So this is an empty range.

File: tests/integrationtest/t/executor/explain.test [hunk 2]
```diff
@@ -214,3 +214,12 @@ create table t (a int, b int, index (a));
...
+explain format='brief' SELECT /*+ AGG_TO_COP() STREAM_AGG()*/ (NOT (`t`.`i`>=_UTF8MB4'j筧8') OR NOT (`t`.`i`=_UTF8MB4'暈lH忧ll6')) IS TRUE,MAX(`t`.`e`) AS `r0`,QUOTE(`t`.`i`) AS `r1` FROM `t` WHERE `t`.`h`>240817 OR `t`.`i` BETWEEN _UTF8MB4'WVz' AND _UTF8MB4'G#駧褉ZC領*lov' GROUP BY `t`.`i`;
+select count(*) from (SELECT /*+ AGG_TO_COP() STREAM_AGG()*/ (NOT (`t`.`i`>=_UTF8MB4'j筧8') OR NOT (`t`.`i`=_UTF8MB4'暈lH忧ll6')) IS TRUE,MAX(`t`.`e`) AS `r0`,QUOTE(`t`.`i`) AS `r1` FROM `t` WHERE `t`.`h`>240817 OR `t`.`i` BETWEEN _UTF8MB4'WVz' AND _UTF8MB4'G#駧褉ZC領*lov' GROUP BY `t`.`i`) derived;
+explain format='brief' SELECT /*+ AGG_TO_COP() */ (NOT (`t`.`i`>=_UTF8MB4'j筧8') OR NOT (`t`.`i`=_UTF8MB4'暈lH忧ll6')) IS TRUE,MAX(`t`.`e`) AS `r0`,QUOTE(`t`.`i`) AS `r1` FROM `t` WHERE `t`.`h`>240817 OR `t`.`i` BETWEEN _UTF8MB4'WVz' AND _UTF8MB4'G#駧褉ZC領*lov' GROUP BY `t`.`i`;
```
Comment (hi-rustin) [inline tests/integrationtest/t/executor/explain.test:224]:
```sql
mysql> select 1 where _UTF8MB4'WVz' > _UTF8MB4'G#駧褉ZC領*lov';
+---+
| 1 |
+---+
| 1 |
+---+
1 row in set (0.00 sec)

mysql> select 1 where _UTF8MB4'WVz' < _UTF8MB4'G#駧褉ZC領*lov';
Empty set (0.00 sec)
```

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check (PR reviewed by myself and ready for feedback.)

File: pkg/util/ranger/ranger_test.go [hunk 3]
```diff
@@ -2290,3 +2290,43 @@ create table t(
 		require.Equal(t, tt.resultStr, got, fmt.Sprintf("different for expr %s", tt.exprStr))
 	}
 }
+
+func TestIssue40997(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	_, err := tk.Exec(`
```
Comment (hawkingrei) [inline pkg/util/ranger/ranger_test.go:2298]:
```suggestion
	err := tk.ExecToErr(`
```

**Review by hawkingrei — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Pull request merged**